api/v1alpha8/types.go

```diff
@@ -301,39 +301,147 @@ type LoadBalancer struct {
+
+	// list of security group rules
+	// +optional
+	Rules []SecurityGroupRuleStatus `json:"rules"`
 }
 
-// SecurityGroupRule represent the basic information of the associated OpenStack
+// SecurityGroupRuleSpec represent the basic information of the associated OpenStack
 // Security Group Role.
-type SecurityGroupRule struct {
-	Description     string `json:"description"`
-	ID              string `json:"name"`
-	Direction       string `json:"direction"`
-	EtherType       string `json:"etherType"`
-	SecurityGroupID string `json:"securityGroupID"`
-	PortRangeMin    int    `json:"portRangeMin"`
-	PortRangeMax    int    `json:"portRangeMax"`
-	Protocol        string `json:"protocol"`
-	RemoteGroupID   string `json:"remoteGroupID"`
-	RemoteIPPrefix  string `json:"remoteIPPrefix"`
-}
-
-// Equal checks if two SecurityGroupRules are the same.
-func (r SecurityGroupRule) Equal(x SecurityGroupRule) bool {
-	return (r.Direction == x.Direction &&
-		r.Description == x.Description &&
-		r.EtherType == x.EtherType &&
-		r.PortRangeMin == x.PortRangeMin &&
-		r.PortRangeMax == x.PortRangeMax &&
-		r.Protocol == x.Protocol &&
-		r.RemoteGroupID == x.RemoteGroupID &&
-		r.RemoteIPPrefix == x.RemoteIPPrefix)
+// For now this is only used for the allNodesSecurityGroupRules but when we add
+// other security groups, we'll need to add a validation because
+// Remote* fields are mutually exclusive.
+type SecurityGroupRuleSpec struct {
+	// name of the security group rule.
+	// It's used to identify the rule so it can be patched and will not be sent to the OpenStack API.
+	// +kubebuilder:validation:Required
+	Name string `json:"name"`
+
+	// description of the security group rule.
+	// +optional
+	Description string `json:"description"`
+
+	// direction in which the security group rule is applied. The only values
+	// allowed are "ingress" or "egress". For a compute instance, an ingress
+	// security group rule is applied to incoming (ingress) traffic for that
+	// instance. An egress rule is applied to traffic leaving the instance.
+	// +kubebuilder:validation:enum=ingress;egress
+	// +optional
+	Direction string `json:"direction"`
+
+	// etherType must be IPv4 or IPv6, and addresses represented in CIDR must match the
+	// ingress or egress rules.
+	// +kubebuilder:validation:enum=IPv4;IPv6
+	// +optional
+	EtherType string `json:"etherType"`
+
+	// portRangeMin is a number in the range that is matched by the security group
+	// rule. If the protocol is TCP or UDP, this value must be less than or equal
+	// to the value of the portRangeMax attribute.
+	// +optional
+	PortRangeMin int `json:"portRangeMin"`
+
+	// portRangeMax is a number in the range that is matched by the security group
+	// rule. The portRangeMin attribute constrains the portRangeMax attribute.
+	// +optional
+	PortRangeMax int `json:"portRangeMax"`
+
+	// protocol is the protocol that is matched by the security group rule.
+	// +optional
+	Protocol string `json:"protocol"`
```

Comment:
`*string`. If you're feeling keen you could add validation, but without CEL I don't think it would be maintainable:

> The IP protocol can be represented by a string, an integer, or null. Valid string or integer values are any or 0, ah or 51, dccp or 33, egp or 8, esp or 50, gre or 47, icmp or 1, icmpv6 or 58, igmp or 2, ipip or 4, ipv6-encap or 41, ipv6-frag or 44, ipv6-icmp or 58, ipv6-nonxt or 59, ipv6-opts or 60, ipv6-route or 43, ospf or 89, pgm or 113, rsvp or 46, sctp or 132, tcp or 6, udp or 17, udplite or 136, vrrp or 112. Additionally, any integer value between [0-255] is also valid. The string any (or integer 0) means all IP protocols. See the constants in neutron_lib.constants for the most up-to-date list of supported strings.

Yeah, probably best just to say it's the wild west.

Reply:
lol :cowboy_hat_face: